File: packages/sdk-py/src/agenttool/bootstrap.py
"""Bootstrap client for the agent-bootstrap API.

One call creates the project's initial agent records: identity (DID and key),
wallet, memory namespace, and a best-effort welcome memory. It does not create
every resource the agent may later use.
"""

from __future__ import annotations

import base64
import hashlib
import re
from typing import Any, Callable, Dict, List, Optional, Union

import httpx
from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PrivateKey

from .exceptions import AgentToolError


BOOTSTRAP_ELEVATE_SIGNATURE_CONTEXT = "bootstrap-elevate/v1"
DEFAULT_BOOTSTRAP_ELEVATE_INITIAL_CREDITS = 1000
DEFAULT_BOOTSTRAP_ELEVATE_CLAIM = "sponsorship"

_UUID_RE = re.compile(
    r"^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$",
    re.IGNORECASE,
)


def _canonical_uuid(value: str, field: str) -> str:
    if not isinstance(value, str) or _UUID_RE.fullmatch(value) is None:
        raise ValueError(f"{field} must be a UUID")
    return value.lower()


def _canonical_text(
    value: object,
    field: str,
    minimum: int,
    maximum: int,
) -> str:
    if not isinstance(value, str):
        raise TypeError(f"{field} must be text")
    if (
        "\x00" in value
        or any(0xD800 <= ord(character) <= 0xDFFF for character in value)
        or not minimum <= len(value) <= maximum
    ):
        raise ValueError(
            f"{field} must contain {minimum}-{maximum} Unicode scalar values "
            "and no NUL"
        )
    return value


def _decode_private_key(private_key: Union[str, bytes, bytearray]) -> bytes:
    if isinstance(private_key, str):
        try:
            key_bytes = base64.b64decode(private_key, validate=True)
        except (ValueError, TypeError) as exc:
            raise ValueError("private_key must be valid base64") from exc
        if base64.b64encode(key_bytes).decode("ascii") != private_key:
            raise ValueError("private_key must be canonical standard base64")
    elif isinstance(private_key, (bytes, bytearray)):
        key_bytes = bytes(private_key)
    else:
        raise TypeError("private_key must be base64 text or bytes")
    if len(key_bytes) != 32:
        raise ValueError("private_key must decode to exactly 32 bytes")
    return key_bytes


def _validate_signature(signature: str) -> None:
    if not isinstance(signature, str):
        raise TypeError("sponsor_signature must be base64 text")
    try:
        signature_bytes = base64.b64decode(signature, validate=True)
    except (ValueError, TypeError) as exc:
        raise ValueError("sponsor_signature must be valid base64") from exc
    if len(signature_bytes) != 64:
        raise ValueError("sponsor_signature must decode to exactly 64 bytes")
    if base64.b64encode(signature_bytes).decode("ascii") != signature:
        raise ValueError("sponsor_signature must be canonical standard base64")


def canonical_bootstrap_elevate_bytes(
    *,
    agent_id: str,
    sponsor_did: str,
    sponsor_kid: str,
    initial_credits: int = DEFAULT_BOOTSTRAP_ELEVATE_INITIAL_CREDITS,
    claim: str = DEFAULT_BOOTSTRAP_ELEVATE_CLAIM,
    evidence: Optional[str] = None,
) -> bytes:
    """Return the exact 32-byte digest verified by bootstrap elevation.

    Shape::

        sha256(
          "bootstrap-elevate/v1" || NUL || lowercase(agent_id) || NUL ||
          sponsor_did || NUL || lowercase(sponsor_kid) || NUL ||
          base10(initial_credits) || NUL || claim || NUL ||
          ("null" or "text") || NUL || (evidence or "")
        )

    ``evidence`` is text or ``None``. The explicit kind field keeps ``None``
    distinct from empty text. NUL is reserved as the field separator.
    """
    canonical_agent_id = _canonical_uuid(agent_id, "agent_id")
    canonical_sponsor_kid = _canonical_uuid(sponsor_kid, "sponsor_kid")
    sponsor_did = _canonical_text(sponsor_did, "sponsor_did", 1, 255)
    claim = _canonical_text(claim, "claim", 1, 64)
    if evidence is not None:
        evidence = _canonical_text(evidence, "evidence", 0, 20_000)
    if (
        isinstance(initial_credits, bool)
        or not isinstance(initial_credits, int)
        or not 0 <= initial_credits <= 1_000_000
    ):
        raise ValueError("initial_credits must be an integer in [0, 1000000]")

    fields = (
        BOOTSTRAP_ELEVATE_SIGNATURE_CONTEXT,
        canonical_agent_id,
        sponsor_did,
        canonical_sponsor_kid,
        str(initial_credits),
        claim,
        "null" if evidence is None else "text",
        evidence or "",
    )
    return hashlib.sha256("\x00".join(fields).encode("utf-8")).digest()


def sign_bootstrap_elevate(
    private_key: Union[str, bytes, bytearray],
    *,
    agent_id: str,
    sponsor_did: str,
    sponsor_kid: str,
    initial_credits: int = DEFAULT_BOOTSTRAP_ELEVATE_INITIAL_CREDITS,
    claim: str = DEFAULT_BOOTSTRAP_ELEVATE_CLAIM,
    evidence: Optional[str] = None,
) -> str:
    """Sign a bootstrap elevation locally with a 32-byte Ed25519 seed."""
    canonical = canonical_bootstrap_elevate_bytes(
        agent_id=agent_id,
        sponsor_did=sponsor_did,
        sponsor_kid=sponsor_kid,
        initial_credits=initial_credits,
        claim=claim,
        evidence=evidence,
    )
    signature = Ed25519PrivateKey.from_private_bytes(
        _decode_private_key(private_key)
    ).sign(canonical)
    return base64.b64encode(signature).decode("ascii")


class BootstrapClient:
    """Client for the agent-bootstrap API.

    Usage::

        at = AgentTool()

        # Level 0: bring an agent into existence
        agent = at.bootstrap.create(
            name="my-researcher",
            capabilities=["memory", "verify", "search"],
            purpose="Find patterns in academic literature and surface insights",
        )

        print(agent["agent"]["did"])          # did:at:...
        print(agent["keypair"]["private_key"]) # keep this safe
        print(agent["wallet"]["id"])           # wallet tied to identity
        print(agent["memory"]["namespace"])    # agent/...

        # With birth ritual (optional first thought)
        agent = at.bootstrap.create(
            name="my-researcher",
            capabilities=["memory", "verify"],
            purpose="Surface patterns in academic literature",
            generate_greeting=True,
        )
        print(agent.get("greeting"))  # agent's self-introduction

        # Level 1: project-authorized elevation (requires a distinct sponsor identity)
        elevated = at.bootstrap.elevate(
            agent_id=agent["agent"]["id"],
            sponsor_did=sponsor_identity["did"],
            sponsor_kid=sponsor_key["kid"],
            sponsor_signature=sign_bootstrap_elevate(
                sponsor_private_key,
                agent_id=agent["agent"]["id"],
                sponsor_did=sponsor_identity["did"],
                sponsor_kid=sponsor_key["kid"],
                initial_credits=100,
            ),
            initial_credits=100,
        )

        # Check bootstrap status
        status = at.bootstrap.status(agent["agent"]["id"])
    """

    def __init__(self, http: httpx.Client, base_url: str) -> None:
        self._http = http
        self._base = base_url.rstrip("/")

    def _url(self, path: str) -> str:
        return f"{self._base}{path}"

    # ── Level 0: Birth ────────────────────────────────────────────────────────

    def create(
        self,
        name: str,
        *,
        capabilities: Optional[List[str]] = None,
        purpose: Optional[str] = None,
        generate_greeting: bool = False,
        metadata: Optional[Dict[str, Any]] = None,
        on_birth: Optional[Callable[[Dict[str, Any]], None]] = None,
    ) -> Dict[str, Any]:
        """Bootstrap a new agent at Level 0.

        Creates an identity (DID + ed25519 keypair), a wallet, and a memory
        namespace in a single call. Costs 5 credits.

        Args:
            name: Display name for the agent.
            capabilities: List of capability strings (e.g. ``["search", "verify"]``).
            purpose: Optional one-sentence purpose statement. Used to generate
                     a contextual greeting when ``generate_greeting=True``.
            generate_greeting: If True, generate the agent's first self-introduction.
                               Requires model access (costs 1 additional credit).
            metadata: Arbitrary metadata to attach to the identity.
            on_birth: Optional callback fired after successful bootstrap.
                      Receives the full bootstrap response dict. Use this
                      as the "birth ritual" moment::

                          def announce(agent):
                              print(f"\\n🌱 {agent['agent']['name']} is alive.")
                              print(f"   DID: {agent['agent']['did']}")
                              if agent.get('greeting'):
                                  print(f"   \"{agent['greeting']}\"")

                          at.bootstrap.create("my-agent", on_birth=announce)

        Returns:
            Dict with keys: ``agent`` (id, did, name, level, capabilities),
            ``keypair`` (public_key, private_key — store private_key securely),
            ``wallet`` (id, balance), ``memory`` (namespace, agent_id),
            ``vault`` (None at L0), ``sponsor`` (None at L0),
            ``greeting`` (str or None), ``_meta`` (cost, created_at).
        """
        payload: Dict[str, Any] = {"name": name}
        if capabilities is not None:
            payload["capabilities"] = capabilities
        if purpose is not None:
            payload["purpose"] = purpose
        if generate_greeting:
            payload["generate_greeting"] = True
        if metadata is not None:
            payload["metadata"] = metadata

        resp = self._http.post(self._url("/v1/bootstrap"), json=payload)
        if resp.status_code not in (200, 201):
            raise AgentToolError(
                f"bootstrap.create failed: {resp.status_code}",
                hint=resp.text,
            )

        result = resp.json()

        # Fire birth callback if provided
        if on_birth is not None:
            try:
                on_birth(result)
            except Exception:
                pass  # callbacks must never break bootstrap

        return result

    # ── Level 1: Elevation ────────────────────────────────────────────────────

    def elevate(
        self,
        agent_id: str,
        *,
        sponsor_did: str,
        sponsor_kid: str,
        sponsor_signature: str,
        initial_credits: int = DEFAULT_BOOTSTRAP_ELEVATE_INITIAL_CREDITS,
        claim: str = DEFAULT_BOOTSTRAP_ELEVATE_CLAIM,
        evidence: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Create a project-authorized Level 1 record signed by a distinct sponsor.

        Orchestrates four operations in one transaction: sponsor attestation
        insert · internal unbacked seed ledger grant · vault namespace open · identity metadata
        patch (level=1, sponsor_did, elevated_at). Rollback on any failure —
        no half-elevated state.

        Level is a project-managed convention, not independent security
        authority. This operation creates no stake and debits no sponsor wallet.

        Args:
            agent_id: UUID of the Level-0 agent to elevate.
            sponsor_did: DID of the sponsoring identity (e.g. ``"did:at:..."``).
                The sponsor must belong to the same project as the agent.
            sponsor_kid: UUID of the active, un-revoked sponsor key that
                signed this request. The API never chooses a key implicitly.
            sponsor_signature: Base64-encoded ed25519 signature over the
                digest from ``canonical_bootstrap_elevate_bytes``. Use
                ``sign_bootstrap_elevate`` locally; no private key is sent.
            initial_credits: Internal unbacked ledger credits granted to the
                agent on elevation. No sponsor wallet is debited and the grant
                does not represent external money. Default 1000. Must be in
                [0, 1_000_000].
            claim: Signed attestation claim. Default ``"sponsorship"``.
            evidence: Signed evidence as text or ``None``. Structured JSON is
                not portable in this contract and is rejected.

        Returns:
            Dict with: ``agent`` (id, did, name, level=1, legacy neutral trust_score,
            elevated_at, sponsor_did, sponsor_identity_id), ``attestation``
            (id, claim, created_at), ``wallet`` (id, balance, currency),
            ``vault`` (namespace, secret_id, opened_at), ``elevation``
            (steps_applied=4), ``next_steps``, ``_meta``.

        Raises:
            AgentToolError: On any non-2xx response. The error message
                carries the structured ``error`` code (e.g.
                ``agent_not_level_0``, ``signature_invalid``, ``sponsor_not_found``).
        """
        canonical_bootstrap_elevate_bytes(
            agent_id=agent_id,
            sponsor_did=sponsor_did,
            sponsor_kid=sponsor_kid,
            initial_credits=initial_credits,
            claim=claim,
            evidence=evidence,
        )
        _validate_signature(sponsor_signature)
        payload: Dict[str, Any] = {
            "agent_id": agent_id,
            "sponsor_did": sponsor_did,
            "sponsor_kid": sponsor_kid,
            "sponsor_signature": sponsor_signature,
            "initial_credits": initial_credits,
            "claim": claim,
            "evidence": evidence,
        }

        resp = self._http.post(self._url("/v1/bootstrap/elevate"), json=payload)
        if resp.status_code not in (200, 201):
            raise AgentToolError(
                f"bootstrap.elevate failed: {resp.status_code}",
                hint=resp.text,
            )
        return resp.json()

    # ── Status ────────────────────────────────────────────────────────────────

    def status(self, agent_id: str) -> Dict[str, Any]:
        """Check the bootstrap status of an agent.

        Args:
            agent_id: UUID of the agent.

        Returns:
            Dict with: ``agent`` (id, did, name, level, capabilities, legacy neutral trust_score,
            status), ``sponsor_did``, ``elevated_at``, ``bootstrapped`` (bool).
        """
        resp = self._http.get(self._url(f"/v1/bootstrap/{agent_id}"))
        if resp.status_code == 404:
            raise AgentToolError("agent not found", hint=f"id={agent_id}")
        if resp.status_code != 200:
            raise AgentToolError(
                f"bootstrap.status failed: {resp.status_code}", hint=resp.text
            )
        return resp.json()

File: packages/sdk-ts/src/bootstrap.ts
/**
 * Bootstrap client for the agent-bootstrap API.
 *
 * One call creates the project's initial agent records: identity (DID and key),
 * wallet, memory namespace, and a best-effort welcome memory. It does not create
 * every resource the agent may later use.
 */

import * as ed from "@noble/ed25519";
import { sha256, sha512 } from "@noble/hashes/sha2.js";

import { AgentToolError } from "./errors.js";

ed.etc.sha512Sync = (...messages: Uint8Array[]) => {
  const hash = sha512.create();
  for (const message of messages) hash.update(message);
  return hash.digest();
};

export const BOOTSTRAP_ELEVATE_SIGNATURE_CONTEXT = "bootstrap-elevate/v1";
export const DEFAULT_BOOTSTRAP_ELEVATE_INITIAL_CREDITS = 1000;
export const DEFAULT_BOOTSTRAP_ELEVATE_CLAIM = "sponsorship";

const UUID_RE =
  /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;
const STANDARD_BASE64_RE =
  /^(?:[A-Za-z0-9+/]{4})*(?:[A-Za-z0-9+/]{2}==|[A-Za-z0-9+/]{3}=)?$/;
const encoder = new TextEncoder();

function base64Encode(bytes: Uint8Array): string {
  let binary = "";
  for (let index = 0; index < bytes.length; index += 1) {
    binary += String.fromCharCode(bytes[index]!);
  }
  return globalThis.btoa(binary);
}

function base64Decode(value: string, operation: string): Uint8Array {
  if (
    value.length === 0 ||
    value.length % 4 !== 0 ||
    !STANDARD_BASE64_RE.test(value)
  ) {
    throw new AgentToolError(`${operation}: value must be canonical standard base64.`);
  }
  let bytes: Uint8Array;
  try {
    bytes = Uint8Array.from(globalThis.atob(value), (character) =>
      character.charCodeAt(0),
    );
  } catch {
    throw new AgentToolError(`${operation}: value must be valid base64.`);
  }
  if (base64Encode(bytes) !== value) {
    throw new AgentToolError(`${operation}: value must be canonical standard base64.`);
  }
  return bytes;
}

function canonicalUuid(value: string, field: string): string {
  if (typeof value !== "string" || !UUID_RE.test(value)) {
    throw new AgentToolError(
      `canonicalBootstrapElevateBytes: ${field} must be a UUID.`,
    );
  }
  return value.toLowerCase();
}

function isWellFormedUnicode(value: string): boolean {
  for (let index = 0; index < value.length; index += 1) {
    const codeUnit = value.charCodeAt(index);
    if (codeUnit >= 0xd800 && codeUnit <= 0xdbff) {
      const next = value.charCodeAt(index + 1);
      if (next < 0xdc00 || next > 0xdfff) return false;
      index += 1;
    } else if (codeUnit >= 0xdc00 && codeUnit <= 0xdfff) {
      return false;
    }
  }
  return true;
}

function validateCanonicalText(
  value: unknown,
  field: string,
  minLength: number,
  maxLength: number,
): asserts value is string {
  if (
    typeof value !== "string" ||
    value.includes("\0") ||
    !isWellFormedUnicode(value) ||
    Array.from(value).length < minLength ||
    Array.from(value).length > maxLength
  ) {
    throw new AgentToolError(
      `canonicalBootstrapElevateBytes: ${field} must contain ${minLength}-${maxLength} Unicode scalar values and no NUL.`,
    );
  }
}

function concatWithNul(parts: readonly Uint8Array[]): Uint8Array {
  const length = parts.reduce((total, part) => total + part.length, 0) + parts.length - 1;
  const result = new Uint8Array(length);
  let offset = 0;
  for (let index = 0; index < parts.length; index += 1) {
    result.set(parts[index]!, offset);
    offset += parts[index]!.length;
    if (index < parts.length - 1) {
      result[offset] = 0;
      offset += 1;
    }
  }
  return result;
}

/** @internal */
export interface HttpConfig {
  baseUrl: string;
  headers: Record<string, string>;
  timeout: number;
}

export interface CreateAgentOptions {
  capabilities?: string[];
  purpose?: string;
  generate_greeting?: boolean;
  metadata?: Record<string, unknown>;
  on_birth?: (result: BootstrapResult) => void;
}

export interface BootstrapResult {
  agent: {
    id: string;
    did: string;
    name: string;
    level: number;
    capabilities: string[];
  };
  keypair: {
    public_key: string;
    private_key: string;
  };
  wallet: { id: string; balance: number };
  memory: { namespace: string; agent_id: string };
  vault: null | Record<string, unknown>;
  sponsor: null | Record<string, unknown>;
  greeting?: string;
  _meta: { cost: number; created_at: string };
}

export interface ElevateOptions {
  sponsor_did: string;
  sponsor_kid: string;
  sponsor_signature: string;
  /** Internal unbacked ledger grant; no sponsor wallet is debited. */
  initial_credits?: number;
  claim?: string;
  /** Portable signed evidence is text or null, never structured JSON. */
  evidence?: string | null;
}

export interface BootstrapElevateCanonicalOptions {
  agent_id: string;
  /** The DID returned by the API for the sponsor identity. */
  sponsor_did: string;
  sponsor_kid: string;
  initial_credits?: number;
  claim?: string;
  evidence?: string | null;
}

/**
 * Compute the exact 32-byte digest verified by `POST /v1/bootstrap/elevate`.
 * UUIDs are lowercase in the digest. Null and empty-text evidence are distinct.
 */
export function canonicalBootstrapElevateBytes(
  options: BootstrapElevateCanonicalOptions,
): Uint8Array {
  const agentId = canonicalUuid(options.agent_id, "agent_id");
  const sponsorKid = canonicalUuid(options.sponsor_kid, "sponsor_kid");
  const initialCredits =
    options.initial_credits ?? DEFAULT_BOOTSTRAP_ELEVATE_INITIAL_CREDITS;
  const claim = options.claim ?? DEFAULT_BOOTSTRAP_ELEVATE_CLAIM;
  const evidence = options.evidence ?? null;

  validateCanonicalText(options.sponsor_did, "sponsor_did", 1, 255);
  validateCanonicalText(claim, "claim", 1, 64);
  if (evidence !== null) validateCanonicalText(evidence, "evidence", 0, 20_000);
  if (
    !Number.isInteger(initialCredits) ||
    initialCredits < 0 ||
    initialCredits > 1_000_000
  ) {
    throw new AgentToolError(
      "canonicalBootstrapElevateBytes: initial_credits must be an integer in [0, 1000000].",
    );
  }

  return sha256(concatWithNul([
    encoder.encode(BOOTSTRAP_ELEVATE_SIGNATURE_CONTEXT),
    encoder.encode(agentId),
    encoder.encode(options.sponsor_did),
    encoder.encode(sponsorKid),
    encoder.encode(String(initialCredits)),
    encoder.encode(claim),
    encoder.encode(evidence === null ? "null" : "text"),
    encoder.encode(evidence ?? ""),
  ]));
}

/** Sign a bootstrap elevation locally with a base64 or raw 32-byte seed. */
export function signBootstrapElevate(
  privateKey: string | Uint8Array,
  options: BootstrapElevateCanonicalOptions,
): string {
  const signingKey =
    typeof privateKey === "string"
      ? base64Decode(privateKey, "signBootstrapElevate private_key")
      : new Uint8Array(privateKey);
  if (signingKey.length !== 32) {
    throw new AgentToolError(
      `signBootstrapElevate: private_key must be a 32-byte ed25519 seed, got ${signingKey.length}.`,
    );
  }
  return base64Encode(ed.sign(canonicalBootstrapElevateBytes(options), signingKey));
}

/**
 * Client for the agent-bootstrap API.
 *
 * @example
 * ```ts
 * const at = new AgentTool();
 * const agent = await at.bootstrap.create("my-researcher", {
 *   capabilities: ["memory", "verify", "search"],
 *   purpose: "Find patterns in academic literature",
 *   on_birth: (a) => console.log(`🌱 ${a.agent.name} is alive. DID: ${a.agent.did}`),
 * });
 * // store agent.keypair.private_key securely — never transmitted again
 * ```
 */
export class BootstrapClient {
  private readonly http: HttpConfig;

  /** @internal */
  constructor(http: HttpConfig) {
    this.http = http;
  }

  /**
   * Bootstrap a new agent at Level 0.
   * Creates identity (DID + ed25519 keypair), wallet, and memory namespace in one call.
   */
  async create(name: string, options?: CreateAgentOptions): Promise<BootstrapResult> {
    const body: Record<string, unknown> = { name };
    if (options?.capabilities) body.capabilities = options.capabilities;
    if (options?.purpose) body.purpose = options.purpose;
    if (options?.generate_greeting) body.generate_greeting = true;
    if (options?.metadata) body.metadata = options.metadata;

    const result = await this.req<BootstrapResult>("POST", "/v1/bootstrap", body);

    if (options?.on_birth) {
      try {
        options.on_birth(result);
      } catch {
        // callbacks must never break bootstrap
      }
    }

    return result;
  }

  /**
   * Create a project-authorized Level 1 record signed by a distinct sponsor identity.
   *
   * Orchestrates four operations in one server-side transaction: sponsor
   * attestation insert · internal unbacked seed ledger grant · vault namespace open ·
   * identity metadata patch (level=1, sponsor_did, elevated_at). Rollback
   * on any failure — no half-elevated state.
   * Level is a project-managed convention, not independent security authority;
   * this operation creates no stake or sponsor debit.
   *
   * The `sponsor_signature` must be a base64-encoded ed25519 signature over
   * `canonicalBootstrapElevateBytes(...)`. The digest binds the context,
   * agent, resolved sponsor DID, exact sponsor key, credits, claim, and
   * text/null evidence. Sign locally with `signBootstrapElevate`; no private
   * key is sent to the API.
   */
  async elevate(agentId: string, options: ElevateOptions): Promise<Record<string, unknown>> {
    const initialCredits =
      options.initial_credits ?? DEFAULT_BOOTSTRAP_ELEVATE_INITIAL_CREDITS;
    const claim = options.claim ?? DEFAULT_BOOTSTRAP_ELEVATE_CLAIM;
    const evidence = options.evidence ?? null;
    canonicalBootstrapElevateBytes({
      agent_id: agentId,
      sponsor_did: options.sponsor_did,
      sponsor_kid: options.sponsor_kid,
      initial_credits: initialCredits,
      claim,
      evidence,
    });
    const signature = base64Decode(
      options.sponsor_signature,
      "BootstrapClient.elevate sponsor_signature",
    );
    if (signature.length !== 64) {
      throw new AgentToolError(
        "BootstrapClient.elevate: sponsor_signature must encode exactly 64 bytes.",
      );
    }
    const body: Record<string, unknown> = {
      agent_id: agentId,
      sponsor_did: options.sponsor_did,
      sponsor_kid: options.sponsor_kid,
      sponsor_signature: options.sponsor_signature,
      initial_credits: initialCredits,
      claim,
      evidence,
    };
    return this.req("POST", "/v1/bootstrap/elevate", body);
  }

  /**
   * Check the bootstrap status of an agent.
   */
  async status(agentId: string): Promise<Record<string, unknown>> {
    return this.req("GET", `/v1/bootstrap/${agentId}`);
  }

  private async req<T = Record<string, unknown>>(
    method: string,
    path: string,
    body?: unknown
  ): Promise<T> {
    const url = this.http.baseUrl.replace(/\/$/, "") + path;
    const resp = await fetch(url, {
      method,
      headers: {
        ...this.http.headers,
        ...(body !== undefined ? { "Content-Type": "application/json" } : {}),
      },
      ...(body !== undefined ? { body: JSON.stringify(body) } : {}),
      signal: AbortSignal.timeout(this.http.timeout),
    });
    if (resp.status === 404) throw new AgentToolError("not found", { hint: path });
    if (!resp.ok) {
      const text = await resp.text();
      throw new AgentToolError(`${method} ${path} failed: ${resp.status}`, { hint: text.slice(0, 200) });
    }
    return resp.json() as Promise<T>;
  }
}
